Analysis/src/GenParticle.cpp

```diff
@@ -0,0 +1,216 @@
+/*! Definiton of analysis::FlatEventInfo class.
+This file is part of https://github.com/hh-italian-group/h-tautau. */
+
+#include "h-tautau/Analysis/include/GenParticle.h"
+
+namespace analysis {
+
+GenParticle::GenParticle(const ntuple::Event& events, size_t n)
+    : index(n)
+{
+    if(n >= events.genParticles_p4.size())
+        throw std::runtime_error("events index is out of range.");
+
+    pdg = events.genParticles_pdg.at(n);
+    status = events.genParticles_status.at(n);
+    genStatusFlags = GenStatusFlags(events.genParticles_statusFlags.at(n));
+    momentum = events.genParticles_p4.at(n);
+}
+
+
+GenEvent::GenEvent(const ntuple::Event& event)
+{
+    genParticles.reserve(event.genParticles_p4.size());
+    for(size_t n = 0; n < event.genParticles_p4.size(); ++n)
+        genParticles.emplace_back(event,n);
+
+    for (size_t relIndex = 0; relIndex < event.genParticles_rel_mIndex.size(); relIndex++) {
+        int motherIndex = event.genParticles_rel_mIndex.at(relIndex);
+        int daughterIndex = event.genParticles_rel_pIndex.at(relIndex);
+
+        GenParticle* mother = &genParticles.at(static_cast<size_t>(motherIndex));
+        GenParticle* daughter = &genParticles.at(static_cast<size_t>(daughterIndex));
+
+        mother->daughters.push_back(daughter);
+        daughter->mothers.push_back(mother);
+    }
+
+    for(const GenParticle& genParticle : genParticles ){
+        if (!genParticle.mothers.size())
+            primaryParticles.insert(&genParticle);
+
+
+         if (genParticle.genStatusFlags.isPrompt())
+            particleCodeMap[std::abs(genParticle.pdg)].insert(&genParticle);
+    }
+}
+
+GenParticleSet GenEvent::GetParticles(int particle_pgd) const
+{
+    GenParticleSet results;
+    const ParticleCodeMap::const_iterator code_iter = particleCodeMap.find(particle_pgd);
+
+    if (code_iter != particleCodeMap.end()){
+        for (const GenParticle* particle : code_iter->second){
+            if(!particle->genStatusFlags.isLastCopy()) continue;
+            results.insert(particle);
+        }
+    }
+    return results;
+}
+
+std::vector<const GenParticle*> GenEvent::GetTypesParticles(std::vector<std::string> type_names, const GenParticle* possible_mother ) const
+{
+    std::vector<int> selected_particles_type_pgd;
```

Comment:
the current implementation would be very slow. Better to start the search from the mother and go down over the tree.